source/common/access_log/access_log_formatter.h

```diff
@@ -95,14 +95,34 @@ class FormatterImpl : public Formatter {
                      const StreamInfo::StreamInfo& stream_info) const override;
 
 private:
-  std::vector<FormatterPtr> formatters_;
+  std::vector<FormatterProviderPtr> providers_;
+};
+
+class JsonFormatterImpl : public Formatter {
+public:
+  JsonFormatterImpl(const std::map<const std::string, const std::string>& format_mapping);
+
+  // Formatter::format
+  std::string format(const Http::HeaderMap& request_headers,
+                     const Http::HeaderMap& response_headers,
+                     const Http::HeaderMap& response_trailers,
+                     const StreamInfo::StreamInfo& stream_info) const override;
+
+  std::map<std::string, std::string> to_map(const Http::HeaderMap& request_headers,
```

Comment:
Does this need to be public? Should it just be a helper function in the `.cc` anonymous namespace?

Reply:
Yeah not too happy about this either, the reasons were:
1. it's kinda hard to test this otherwise. exposing the map function makes it easy to just compare dicts instead of going through a parse/unparse dance. this isn't a super strong reason though
2. If we ever support other structured log formats (like proto), then the idea is that we would just rely on this function to create a c++ map, and then each new format would be a simple "serializer" that writes that out (even the current json stuff can be thought of like that). In that case, testing the map function provides more peace of mind

Reply:
If you feel strongly though, lmk and I'm happy to make it private